Interpreter - Add transaction context

ratik21 opened the pull request:
- Added context for transaction using `mkTransaction`
- Added types for `Transaction` (`fields`, `type`, `oncomplete`), `Assets` (`holdings`, `params`) 

NOTE: Will be adding context for state related data with stateful opcodes.

amityadav0 commented on packages/algorand-js/src/types.ts:
```diff
@@ -11,6 +11,109 @@ export type AppArgs = Array<string | number>;
 export type TEALStack = IStack<bigint | Uint8Array>;
 
+// https://developer.algorand.org/docs/reference/teal/opcodes/#txn
+export enum TxnField {
```
I think we already have transaction field in algosdk types. Let's use it.

ratik21 replied:
- This is an `enum` not an interface (are enums present in types/algosdk)?
- I couldn't find an interface also. Can you be more specific?

amityadav0 replied:
how are we planning to use this enum in context ? 
`mkTransaction` will return `transaction` interface.
User cannot pass enum, mkTransaction will not pass enum
how `txn` field will return required value ?

you can search `Transaction`. there is a interface with this name in index.d.ts in types@algosdk.

ratik21 replied:
We are already using `Transaction` interface in the `createContext` function. 
`TxnField` enum is used to represent the transaction fields. (you can check txn opcode spec)
Use Case:
 - TEAL:
```
...
txn Fee
...
```
- PARSER:
```
push `new Txn(TxnField.Fee, this.interpreter)` to logic[]
```
- INTERPRETER 
```
reads this.interpreter.tx and pushes Fee to stack (as the txfield is Fee) 
```  
Let me know if that's clear

amityadav0 replied:
what i am trying to ask here is:
- we get transaction interface from mkTransaction
- interpreter will push Fee to stack
- how are we connecting enum with interface because currently they are disconnected, (you will get blank value for each field)
does this makes sense ?

ratik21 replied:
Yes. We will have to create a function which will return specified value from transaction. This should be done because the TEAL opcode specification for transaction fields are different (in naming) than the `Transaction` interface of sdk. This would be done while adding the txn related opcodes. 
Something like:
```
function getTxnInfo(field: TxnField, transaction: Transaction) {
  switch (TxField) {
   case Fee:
        return transaction.fee
   ... 
   ...
 }
}
```
Should this be handled some other way?

amityadav0 replied:
I was thinking of switching with strings here, this will result in less repetitive code (because we still have to switch with strings in parser)
but we have `TypeEnum`, i think handling with enum is better.

ratik21 replied:
yes, using enums is better than using strings.

amityadav0 commented on packages/algorand-js/src/types.ts:
```diff
@@ -11,6 +11,109 @@ export type AppArgs = Array<string | number>;
+  AssetFrozen // Is the asset frozen or not
+}
+
+export enum AssetParam {
```
let's add a comment here `this is for global storage`
and above type will be mapped to specific account.

robert-zaremba replied:
Yes, we should have docstrings wherever possible

ratik21 replied:
added.

ratik21 commented on packages/algorand-js/src/interpreter/interpreter.ts:
```diff
@@ -12,24 +15,53 @@ export class Interpreter {
   bytecblock: Uint8Array[];
   intcblock: BigInt[];
   scratch: StackElem[];
+  txnContext: Transaction | Transaction[];
```
Ok. BDW, just to confirm: for `n`  transactions as an `atomic` group, there is only single execution of smart contract.

robert-zaremba replied:
Yes

ratik21 commented on packages/algob/src/lib/tx.ts:
```diff
@@ -108,8 +108,7 @@ export function encodeNote (note: string | undefined, noteb64: string| undefined
  */
-async function mkTransaction (deployer: AlgobDeployer, txnParam: execParams): Promise<Transaction> {
-  const params = await mkSuggestedParams(deployer.algodClient, txnParam.payFlags);
+export function mkTransaction (txnParam: execParams, suggestedParams: SuggestedParams): Transaction {
```
I wanted to remove the deployer from `mkTransaction`. So that i can reuse in interpreter (without deployer client)

robert-zaremba replied:
ok

amityadav0 commented on packages/algorand-js/src/types.ts:
```diff
@@ -11,6 +11,109 @@ export type AppArgs = Array<string | number>;
+  DeleteApplication
+}
+
+export enum GlobalField {
```
I don't think we are using this enum in code(same for others). Is this for future use or i missed it ?
can you please elaborate how are we planning to use this ?

ratik21 replied:
yes, sorry. This will be used for `global` opcode.

amityadav0 reviewed (commented):
are we storing values of ASA, balances for a specific account ?

ratik21 commented:
> are we storing values of ASA, balances for a specific account ?

This Pull request is for transaction context. I added a note in the description also.
Anyway, i have added ctx for state related data as well. Can you please check?

robert-zaremba commented on packages/algorand-js/src/interpreter/interpreter.ts:
```diff
@@ -1,35 +1,85 @@
+      const mockParams = mockSuggestedParams(txnParams.payFlags);
+      this.tx = mkTransaction(txnParams, mockParams);
+      this.gtxs = [this.tx];
+    }
```
is the `else` part for tests? 
Hacking a normal code for tests execution is usually a pattern. Why not passing mock params through an argument to this function?

ratik21 replied:
The `else` part is not for tests. It represents a single transaction (not a transaction group).
`mock Params` are not passed explicitly because some params are dynamic (eg `txn fee`). 
And for an atomic group, `mock Params` might be different for different transactions (eg - different `fee` / `firstValid` /..etc)

ratik21 replied:
If the `txnParams` is an Array then it is representing a transaction group (Multiple transactions as an atomic group). So each object in that array represents a single transaction.

Otherwise (the `else` part), we create and assign a single transaction.

robert-zaremba replied:
I meant to include this explanation in the code next time you will do a PR, ;)

robert-zaremba reviewed (approved):
Pre-approving, but please add testing hacks to the  "production" code
